joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -191,39 +197,41 @@ controller_interface::return_type JointTrajectoryController::update(
       const bool before_last_point = end_segment_itr != (*traj_point_active_ptr_)->end();
 
       // set values for next hardware write()
+      if (use_closed_loop_pid_adapter)
+      {
+        // Update PIDs
+        for (auto i = 0ul; i < joint_num; ++i)
+        {
+          tmp_command_[i] = (state_desired.velocities[i] * ff_velocity_scale_[i]) +
+                            pids_[i]->computeCommand(
+                              state_desired.positions[i] - state_current.positions[i],
+                              state_desired.velocities[i] - state_current.velocities[i],
+                              (uint64_t)period.nanoseconds());
+        }
+      }
       if (has_position_command_interface_)
       {
         assign_interface_from_point(joint_command_interface_[0], state_desired.positions);
       }
       if (has_velocity_command_interface_)
       {
-        if (!use_closed_loop_pid_adapter)
+        if (use_closed_loop_pid_adapter)
         {
-          assign_interface_from_point(joint_command_interface_[1], state_desired.velocities);
+          assign_interface_from_point(joint_command_interface_[1], tmp_command_);
         }
         else
         {
-          // Update PIDs
-          for (auto i = 0ul; i < joint_num; ++i)
-          {
-            tmp_command_[i] = (state_desired.velocities[i] * ff_velocity_scale_[i]) +
-                              pids_[i]->computeCommand(
-                                state_desired.positions[i] - state_current.positions[i],
-                                state_desired.velocities[i] - state_current.velocities[i],
-                                (uint64_t)period.nanoseconds());
-          }
-          assign_interface_from_point(joint_command_interface_[1], tmp_command_);
+          assign_interface_from_point(joint_command_interface_[1], state_desired.velocities);
         }
       }
       if (has_acceleration_command_interface_)
       {
         assign_interface_from_point(joint_command_interface_[2], state_desired.accelerations);
       }
-      // TODO(anyone): Add here "if using_closed_loop_hw_interface_adapter" (see ROS1) - #171
-      //       if (check_if_interface_type_exist(
-      //           command_interface_types_, hardware_interface::HW_IF_EFFORT)) {
-      //         assign_interface_from_point(joint_command_interface_[3], state_desired.effort);
-      //       }
+      if (has_effort_command_interface_ && use_closed_loop_pid_adapter)
+      {
+        assign_interface_from_point(joint_command_interface_[3], tmp_command_);
```

Comment:
Should we use here the same logic as for velocity? To assign efforts directly if no pids are used?

Reply:
Actually, would it make sense to directly set the effort or use it without PIDs? Velocity is a physical quantity, so I can see why you can set it directly. However, as we discussed and concluded before, effort is an abstract quantity, which I think means that it's not possible to directly set it. It's only used indirectly to modify the physical quantities position, velocity, and acceleration via the PID.

Reply:
I don't know about that. Exactly because it is an abstract quantity, it could be “anything”. On the other hand, you are right because we would need another interpolator to see it directly. So until we have one… we can keep it also as it is.